wpilib/src/spi.rs

```diff
@@ -95,70 +98,112 @@ impl Spi {
         hal_call!(HAL_SetSPIChipSelectActiveLow(self.port))
     }
 
-    pub fn write(&mut self, buf: &[u8]) -> i32 {
-        unsafe { HAL_WriteSPI(self.port, buf.as_ptr(), buf.len() as i32) }
+    pub fn write(&mut self, data: &[u8]) -> io::Result<usize> {
+        io_result(unsafe { HAL_WriteSPI(self.port, data.as_ptr(), data.len() as _) })
     }
 
-    pub fn read(&mut self, initiate: bool, buf: &mut [u8]) -> i32 {
+    pub fn read(&mut self, initiate: bool, buf: &mut [u8]) -> io::Result<usize> {
         if initiate {
             let send_data: Vec<u8> = vec![0; buf.len()];
-            return unsafe {
-                HAL_TransactionSPI(
-                    self.port,
-                    send_data.as_ptr(),
-                    buf.as_mut_ptr(),
-                    buf.len() as i32,
-                )
-            };
+            return unsafe { self.transaction_into(&send_data, buf.as_mut_ptr()) };
         }
-        unsafe { HAL_ReadSPI(self.port, buf.as_mut_ptr(), buf.len() as i32) }
+
+        io_result(unsafe { HAL_ReadSPI(self.port, buf.as_mut_ptr(), buf.len() as _) })
     }
 
-    /// Caller must ensure that the slices size is greater than the size parameter
-    pub unsafe fn transaction(&mut self, to_send: &[u8], to_receive: &mut [u8], size: i32) -> i32 {
-        HAL_TransactionSPI(self.port, to_send.as_ptr(), to_receive.as_mut_ptr(), size)
+    /// Performs an SPI send/receive transaction.
+    pub fn transaction(&mut self, to_send: &[u8]) -> io::Result<Vec<u8>> {
+        let size = to_send.len();
+        let mut receive_buf = Vec::with_capacity(size);
+        let read = unsafe { self.transaction_into(to_send, receive_buf.as_mut_ptr()) }?;
+        unsafe { receive_buf.set_len(read) }
+        Ok(receive_buf)
     }
 
-    pub fn init_auto(&mut self, buffer_size: i32) -> HalResult<()> {
-        hal_call!(HAL_InitSPIAuto(self.port, buffer_size))
+    pub unsafe fn transaction_into(
+        &mut self,
+        to_send: &[u8],
+        receive_buf: *mut u8,
+    ) -> io::Result<usize> {
+        let size = to_send.len();
+        io_result(HAL_TransactionSPI(
+            self.port,
+            to_send.as_ptr(),
+            receive_buf,
+            size as _,
+        ))
     }
+}
 
-    pub fn free_auto(&mut self) -> HalResult<()> {
-        hal_call!(HAL_FreeSPIAuto(self.port))
+/// Automatic SPI transfer engine.
+#[derive(Debug)]
+pub struct AutoSpi(Spi);
+
+impl AutoSpi {
+    /// Initialize automatic SPI transfer engine.
+    ///
+    /// Only a single engine is available.
+    /// This will error if an engine is currently already allocated.
+    pub fn new(spi: Spi, buffer_size: i32) -> HalResult<Self> {
```

Comment:
This interface does have the unfortunate consequence that if an error does occur, the `Spi` is lost...

Reply:
We could return `Result<Self, (HalError, SPI)>`. Looking at the HAL code, the `Spi` is trivially recoverable for `PARAMETER_OUT_OF_RANGE` and `RESOURCE_IS_ALLOCATED`, but I'm not sure if what's behind `spiSystem->writeAutoChipSelect(port, status);`. Perhaps this is the safer choice.